### ISSUE TITLE
Very high memory usage

### ISSUE THREAD

**Issue by orlandothoeny** [infection/infection#1996]
We run infection on GCP VMs as part of our CI pipeline.
Our VMs have 2 AMD Milan vCPUs. 8 GB memory and a 200GB SSD with a 12k IOPS and 240MiB/s throughput limit. 

We observe high disk read IOPS and read throughput while Infection is running. These read operations completely saturate our disk. The VM is unresponsive while the job is running.
![Screenshot from 2024-07-29 09-58-28](https://github.com/user-attachments/assets/61d27e64-e551-4d24-81e4-b4e02e25ea38)

Is this behavior normal?
If so, what is Infection doing that causes this? Does it read every file from disk each time before it "applies" a mutation to it?
Does somebody have a suggestion how to fix this? Just increase the disk performance? Or does it need more CPU? I also thought about using a ramdisk.

----

We are running Infection with a single thread.

We are using the following versions:
- infection/infection (0.29.6)
- infection/mutator (0.4.0)
- infection/include-interceptor (0.2.5)
- infection/abstract-testframework-adapter (0.5.0)

**Comment by orlandothoeny** [comment]
Just watched the process using htop. It looks like we might not have enough memory.

**Comment by sanmai** [comment]
It sounds like a runaway mutation: that causes an infinite loop. You can work around it by ignoring the mutations that causes this problem:

**Comment by orlandothoeny** [comment]
Hi @sanmai,

Yeah I also thought about setting the memory limit. This would prevent that the VM hangs itself, but would also cause the job to fail and not fix the root cause.

Do you have an idea how I can find out which mutation is problematic?

**Comment by sanmai** [comment]
How would it cause the job to fail? 

Infection controls all workers it starts, so it should be able to control this runaway mutation with grace.

**Comment by theofidry** [comment]
@orlandothoeny could you check how many processes are spawned by infection? There was recently an issue reported by the author of PHPUnit that showed that the underlying library checking the number of cores was picking up wrong numbers (basically picking the limit declared by the VM, VM which was vastly over-committing resources).

Another thought is do you have guarantees that your tests are not doing anything on the FS? If they do try without those peculiar tests. This could be the start of a lead in the sense that a mutation could cause a weird behavour that's not caught early enough by your tests resulting in rogue code being executed for too long (infection does have a timeout setting though so it shouldn't run indefinitely).

**Comment by orlandothoeny** [comment]
@sanmai The PHP process gets killed when it reaches the configured `memory_limit` threshold. Do you mean that Infection has it's own logic for managing memory so that it does not reach the limit (if configured)?

----

@theofidry We have some integration tests that read files from fixed locations, like JSON schema files and ORM fixtures. But I didn't find any that read something that the subject under tests generates.

I ran Infection with a coverage report that does not include the integration tests that read the mentioned files. Resulting in those tests to be skipped by Infection.
The result was the same as before.

We have the `timeout` setting set to 20.

Here's our active mutators, if that helps:

```json
{
    "$schema": "vendor/infection/infection/resources/schema.json",
    "source": {
        "directories": [
            "src"
        ]
    },
    "initialTestsPhpOptions": "-d xdebug.mode=coverage",
    "timeout": 20,
    "logs": {
        "json": "build/logs/infection.json"
    },
    "minCoveredMsi": 62,
    "mutators": {
        "@arithmetic": true,
        "@boolean": true,
        "@cast": true,
        "CastInt": false,
        "@conditional_boundary": true,
        "@conditional_negotiation": true,
        "@equal": true,
        "NotIdenticalNotEqual": false,
        "IdenticalEqual": false,
        "@identical": true,
        "@number": true,
        "@operator": true,
        "@regex": true,
        "@removal": true,
        "@return_value": true,
        "@sort": true,
        "@unwrap": true,
        "@extensions": true,
        "global-ignore": [
            "Psr\\Log\\LoggerInterface"
        ],
        "global-ignoreSourceCodeByRegex": [
            "\\$this->logger->.*",
            "\\$this->io->.*",
            "\\$io->.*",
            ".*writeln.*"
        ]
    }
}
```

**Comment by orlandothoeny** [comment]
I also though the problem might be, that we disable garbage collection in our CLI `php.ini`.
I added a bootstrap file that runs `gc_enable()`, but that also didn't help

**Comment by orlandothoeny** [comment]
**TL;DR;**
Mutation caused recursive method to never exit and run forever.
----
I ran the following commands, running them on my local machine which has more memory than the CI runners:

```shell
# Generate coverage reports
php -d xdebug.mode=coverage bin/phpunit --coverage-xml=infection/coverage-xml --log-junit=infection/junit.xml --disable-coverage-ignore -c ./phpunit.xml
# Run Infection using the generated reports
/usr/bin/time -v bin/infection --threads=1 --skip-initial-tests --only-covering-test-cases --ignore-msi-with-no-mutations --coverage=./infection
```

Peak memory usage: **17GB**

Infection Output:
```

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

#StandWithUkraine

Infection - PHP Mutation Testing Framework version 0.29.6

[warning] Skipping the initial test run can be very dangerous.
It is your responsibility to ensure the tests are in a passing state to begin.
If this is not done then mutations may report as caught when they are not.


Generate mutants...

Processing source code files: 619/619
.: killed, M: escaped, U: uncovered, E: fatal error, X: syntax error, T: timed out, S: skipped, I: ignored

IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII   (  50 / 7642)
### Omitted ###
......M.....M.......M.M...M.......M.M....MMM.MM...   (7600 / 7642)
..M..M.M................MM................           (7642 / 7642)

7642 mutations were generated:
    2743 mutants were killed
    1288 mutants were configured to be ignored
    2829 mutants were not covered by tests
     740 covered mutants were not detected
       0 errors were encountered
       0 syntax errors were encountered
       3 time outs were encountered
      39 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 43%
         Mutation Code Coverage: 55%
         Covered Code MSI: 78%

Generated Reports:
         - /var/www/REDACTED/build/logs/infection.json

Please note that some mutants will inevitably be harmless (i.e. false positives).
 ! [NOTE] The Covered Code MSI is 16.77% percentage points over the required Covered Code MSI. Consider increasing the  
 !        required Covered Code MSI percentage the next time you run Infection.                                         


Time: 8m 52s. Memory: 0.43GB. Threads: 1
        Command being timed: "bin/infection --threads=1 --skip-initial-tests --only-covering-test-cases --ignore-msi-with-no-mutations --coverage=./infection"
        User time (seconds): 362.11
        System time (seconds): 139.45
        Percent of CPU this job got: 94%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 8:53.55
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 17248160
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 25104
        Minor (reclaiming a frame) page faults: 31159989
        Voluntary context switches: 520707
        Involuntary context switches: 27623
        Swaps: 0
        File system inputs: 607968
        File system outputs: 413000
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

The process that used the most memory was a PHPUnit run for a test that has 12 test-data sets provided by a data provider.
The only thing that I see that is special about the test & SUT is that the SUT contains a recurive method.

As expected, the memory usage was also very high (17GB) after running Infection only for that single SUT class (using `--filter`) 
```shell
/usr/bin/time -v bin/infection --threads=1 --skip-initial-tests --only-covering-test-cases --ignore-msi-with-no-mutations --coverage=./infection --filter=src/XXXXX/XXXBuilder.php --show-mutations --debug -vvv
```

Having a look at the applied mutations, there was one that ran into a timeout.
The diff inside the log for the mutation that ran into the timeout showed a change that caused the recursive method to never exit (it exits if conditions are fulfilled, due to the mutation this never happened).

The possible changes I see are:
* Reduce timeout
* Setting `@infection-ignore-all` annotation on the affected method

Is there a way to skip classes that call a recursive method at some point? Probably not?
Or do you have any other ideas that would prevent this from happening in other cases?

A max. memory setting per mutation - similar to `timeout` -  or something along those lines might help to prevent this.

**Comment by maks-rafalko** [comment]
> Or do you have any other ideas that would prevent this from happening in other cases?

we already have logic to set memory limit for mutant processes, see

https://github.com/infection/infection/blob/b003774100ced9afcfd74e4440f2392b0405d3ba/src/Resource/Memory/MemoryLimiter.php#L84-L101

It looks like it doesn't work on your case. This is what I would suggest:

- on you reproducible environment, could you please debug this class and see if it sets memory limit or not. If not - why?
- or if you can create a reproducer in our [e2e tests](https://github.com/infection/infection/tree/master/tests/e2e), that would be nice so we can investigate on our end

**Comment by sanmai** [comment]
> Is there a way to skip classes that call a recursive method at some point? Probably not?

If there's a method that the recursive method calls, you can mock the former and install the invocation count limit. So PHPUnit will prevent the infinite loop by failing the test.

**Comment by orlandothoeny** [comment]
I think I see the problem. We use pre-generated PHPUnit coverage reports.

The MemoryLimiter is never called, because this condition aborts before that: https://github.com/orlandothoeny/infection/blob/b003774100ced9afcfd74e4440f2392b0405d3ba/src/Engine.php#L94-L94

Additionally, this condition would also abort, because we have no console output from PHPUnit:
https://github.com/orlandothoeny/infection/blob/b003774100ced9afcfd74e4440f2392b0405d3ba/src/Resource/Memory/MemoryLimiter.php#L78-L78

**Comment by orlandothoeny** [comment]
I tried to write a test case (https://github.com/orlandothoeny/infection/commit/429ed1e2c598d4e42f1173fd5185704d0068398d) but can't get it to cause an OOM kill. But the strange thing is, the mutation is marked as `Killed`, not `Errored`, which should happen if the memory limit is reached (according to the existing `Memory_Limit` E2E test). As if `$condition` was not mutated to `true`.

**Comment by sanmai** [comment]
You are correct: MemoryLimiter is a convenience feature that is not always triggered. That's why I suggested adding a sensible memory limit: if you don't like the processes expanding to 7GB, then it means the memory limit is set too high.

You can also follow the steps of Infection: run your tests with coverage, note the peak memory consumption, and set the double of that in PHPUnit's bootstrap script.

**Comment by orlandothoeny** [comment]
I am aware that setting the PHP memory limit would fix the crashes. This would however also cause the CI kobs to fail regularly. Which makes the pipeline flaky, reducing its usefulness, because nobody would react to "real" failures anymore. So before doing that, I could just as well delete the CI job altogether.

Running the tests within Infection might be the easiest fix, however not that great either. Since we have some quite large projects and running the tests twice increases the CI runtime by a couple of minutes.

**Comment by sanmai** [comment]
I'm not quite sure I understand how would the PHP memory limit that is twice the usual memory consumption would cause the tests to be flaky. What it would take to elaborate?

**Comment by orlandothoeny** [comment]
Ah now I understand what you mean. I was thinking about setting the memory limit when running `bin/infection`, not via the PHPUnit config. Yeah I think that should work. Thanks for elaborating

**Comment by sanmai** [comment]
@orlandothoeny Any progress so far?

**Comment by orlandothoeny** [comment]
Calling `ini_set('memory_limit', XXX)` inside the PHPUnit bootstrap file works.
I wrote some code that checks how much memory is free on the Container or VM. I use a percentage of that number for the `memory_limit`. This way it should be more robust than setting a hard limit, and also work when the memory usage on a CI runner is high.

I'll see if I can open-source that code as a Composer package, so you can use (or copy the parts that may be useful for you) it if you want to extend the `MemoryLimiter` logic with a fallback for the case where the tests are not run together with Infection.

**Comment by sanmai** [comment]
Thanks for confirming the proposed approach works. I filed a feature request #1998 as a follow-up.
